File: backend/app/schemas/purchase.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class PurchaseProductRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    sku: str
    name: str
    barcode: str | None
    cost_price: Decimal
    stock: int


class PurchaseProductListResponse(BaseModel):
    items: list[PurchaseProductRead]
    total: int


class PurchaseItemLineCreate(BaseModel):
    product_id: int = Field(ge=1)
    quantity: int = Field(ge=1)
    # Ignored by API — unit cost is always taken from product.cost_price (security).
    unit_cost: Decimal | None = Field(default=None, ge=0)


class PurchaseCreate(BaseModel):
    supplier_id: int = Field(ge=1)
    items: list[PurchaseItemLineCreate] = Field(min_length=1)


class PurchaseItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    product_name: str
    product_sku: str
    quantity: int
    unit_cost: Decimal
    line_total: Decimal


class PurchaseRead(BaseModel):
    id: int
    supplier_id: int
    supplier_name: str
    created_at: datetime
    status: str
    procurement_run_id: int | None = None
    agent_summary: str | None = None
    items: list[PurchaseItemRead]
    total: Decimal


class PurchaseListItem(BaseModel):
    id: int
    supplier_id: int
    supplier_name: str
    created_at: datetime
    item_count: int
    total: Decimal
    status: str
    procurement_run_id: int | None = None
    agent_summary: str | None = None


class PurchaseListResponse(BaseModel):
    items: list[PurchaseListItem]
    total: int

File: frontend/src/types/purchase.ts
export type PurchaseProduct = {
  id: number
  sku: string
  name: string
  barcode: string | null
  cost_price: string
  stock: number
}

export type PurchaseProductListResponse = {
  items: PurchaseProduct[]
  total: number
}

export type PurchaseLineDraft = {
  product_id: number
  sku: string
  name: string
  quantity: number
  unit_cost: string
}

export type PurchaseListItem = {
  id: number
  supplier_id: number
  supplier_name: string
  created_at: string
  item_count: number
  total: string
  status: string
  procurement_run_id: number | null
  agent_summary: string | null
}

export type PurchaseListResponse = {
  items: PurchaseListItem[]
  total: number
}

export type PurchaseItemRead = {
  id: number
  product_id: number
  product_name: string
  product_sku: string
  quantity: number
  unit_cost: string
  line_total: string
}

export type PurchaseRead = {
  id: number
  supplier_id: number
  supplier_name: string
  created_at: string
  status: string
  procurement_run_id: number | null
  agent_summary: string | null
  items: PurchaseItemRead[]
  total: string
}

export type ProcurementRunResponse = {
  id: number
  status: string
  draft_purchase_ids: number[]
  warnings: string[]
  error_message: string | null
}
